### PR TITLE
CPU-Strided-Complex Support for reduce ops and linpack ops

### PR DESCRIPTION
**Pull request by dylanbespalko**
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

Changes so far:

- [x]  Renamed references to variable "I" that may be confused for "I" defined in complex.h.  I did this to avoid crazy CI failures messages as complex.h is included by more source files.
     - aten/src/ATen/native/cpu/Loops.h (Renamed I to INDEX)
     - aten/src/ATen/native/cuda/Loops.cuh (Renamed I to INDEX)
     - aten/src/ATen/core/ivalue_inl.h (Renamed I to INDEX)
     - c10/util/Array.h (Renamed I to INDEX)
     - c10/util/C++17.h (Renamed I to INDEX)
    - c10/util/Metaprogramming.h (Renamed I to INDEX)
    - c10/util/SmallVector.h (custom renaming)
- [x]  Added complex support of Linear Algebra Ops.
     - SVD needed to be modified to support mixed data types 
     - Example U(std::complex<double)), S(double), V(std::complex<double>)
     - See before and after benchmark below (No observable change in performance).
- [x]  Added complex support of Reduce Ops.
     - var/std computations could have been faster if it was possible to interpret std::complex<double> Tensor as a double Tensor. 
- [x]  Added complex derivative support for autograd functionality.
     - derivatives are the same as defined by numpy autograd library for real(), imag(), conj(), angle(). These functions only affect complex numbers.
     - derivative of abs() has not been modified to not interfere with existing code.
     - Autograd defines abs() for complex numbers and fabs() for real numbers. I will look into this further down the road.

 ----------------------------------------
 PyTorch/Caffe2 Operator Micro-benchmarks Before Changes
----------------------------------------
Tag : short

Benchmarking PyTorch: svd
Mode: Eager
Name: svd_M512_N512
Input: M: 512, N: 512
Forward Execution Time (us) : 162339.425
Forward Execution Time (us) : 162517.479
Forward Execution Time (us) : 162847.775

----------------------------------------
PyTorch/Caffe2 Operator Micro-benchmarks After Changes
----------------------------------------
Tag : short

Benchmarking PyTorch: svd
Mode: Eager
Name: svd_M512_N512
Input: M: 512, N: 512
Forward Execution Time (us) : 162032.117
Forward Execution Time (us) : 161943.484
Forward Execution Time (us) : 162513.786

### CODE REVIEW

**Comment by dylanbespalko** [comment]
@ezyang, 

I have added complex number support for linear algebra ops and reduce ops. There are a lot of CI failures after pulling changes from master. I am mainly seeing jit failures `test_cpp`, `test_cpp_cuda`. Can you verify that those errors are not related to me.

**I don't mind waiting on this PR if you need to fix some other problems.**

Dylan

**Comment by ezyang** [comment]
Windows failure is real:

```
17:09:37 C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\complex.h(91): error: more than one instance of overloaded function "norm" has "C" linkage
17:09:37 
17:09:37 C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt\complex.h(116): error: more than one instance of overloaded function "normf" has "C" linkage
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(15): error: namespace "std" has no member "complex"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(15): error: expected a ">"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(15): error: expected a ";"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(20): error: namespace "std" has no member "complex"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(20): error: expected a ">"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(20): error: expected a ";"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(30): error: namespace "std" has no member "complex"
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(30): error: type name is not allowed
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(30): error: function template "VALUE_TYPE at::native::<unnamed>::zabs<SCALAR_TYPE,VALUE_TYPE>(SCALAR_TYPE)" is not an entity that can be explicitly specialized
17:09:37 
17:09:37 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/aten/src\ATen/native/cpu/zmath.h(30): error: expected a ";"
17:09:37 
17:09:37 12 errors detected in the compilation of "C:/Windows/TEMP/tmpxft_00000fb8_00000000-10_ReduceOpsKernel.cpp1.ii".
17:09:37 -- Removing C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/./torch_generated_ReduceOpsKernel.cu.obj
17:09:37 "C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/./torch_generated_ReduceOpsKernel.cu.obj
17:09:37 CMake Error at torch_generated_ReduceOpsKernel.cu.obj.Release.cmake:281 (message):
17:09:37   Error generating file
17:09:37   C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/build/caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/./torch_generated_ReduceOpsKernel.cu.obj
17:09:37 
````

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1001,6 +1111,9 @@ static void apply_svd(Tensor& self, Tensor& U, Tensor& S, Tensor& VT,
   auto m = self.size(-2);
   auto n = self.size(-1);
   auto k = std::min(m, n);
+  int64_t lrwork = jobz == 'N' ? 5*k : k*std::max(5*k+7,2*std::max(m,n)+2*k+1);
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: Where did this formula come from?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -765,7 +872,7 @@ static void apply_geqrf(Tensor& self, Tensor& tau, int64_t m, int64_t n,
   int lwork = -1;
   scalar_t wkopt;
   lapackGeqrf<scalar_t>(m, n, self_data, m, tau_data, &wkopt, lwork, &info);
-  lwork = static_cast<int>(wkopt);
+  lwork = static_cast<int>(zabs<scalar_t, value_t>(wkopt));
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: I thought about it a while and couldn't figure it out. Why is the zabs call here necessary?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (dylanbespalko) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Sorry for the confusion. I meant to take this settings from the LAPACK docs, but I think I copied the wrong setting when I was searching other sites for the reasoning behind this setting. The code has been updated with the setting for LAPACK 3.8.

The rwork setting in [LAPACK](http://www.netlib.org/lapack/explore-html/d3/da8/group__complex16_g_esing_gaccb06ed106ce18814ad7069dcb43aa27.html) is the minimum setting. Increasing this number by 2x, 10x, and 100x did not impact performance in an measurable way.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (dylanbespalko) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: wkopt cannot be cast directly to the lwork data type without converting to a real number. wkopt is a positive real number that is stored in a std::complex variable thus `zabs()` and `real_impl()` should both do the job, but `real_impl()` is more performant.

I have updated the template utility functions in `zmath.h` to support:
 - complex return type: when performing tensor operations that maintain a fixed dtype.
 - real return type: when performing low-level C++ operations.
All of these functions are no-ops for other data types.

**Comment by dylanbespalko** [comment]
@ezyang,

I have fixed the windows CI failure by name mangling a couple variable names called `norm` to `norm_` that conflicted with some C references to a function called `norm` that doesn't belong to any namespace. All the CI looks good now.

File: aten/src/ATen/Dispatch.h [hunk 3]
```diff
@@ -313,6 +313,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                                                            \
   }()
 
+#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(SCALARTYPE1, SCALARTYPE2, TYPE, NAME, ...)                        \
```
Comment (ezyang) [inline aten/src/ATen/Dispatch.h]: @gchanan can you remind me what the rules for adding dispatch macros here?

**Review by ezyang — APPROVED** [review]

**Comment by dylanbespalko** [comment]
@ezyang,

Are you able to merge this PR? Thanks.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -1000,8 +1110,19 @@ static void apply_svd(Tensor& self, Tensor& U, Tensor& S, Tensor& VT,
...
+  if (jobz == 'N'){
+    lrwork = 5*mn;
+  }else if (mx > 10*mn){
+    lrwork = 5*mn*mn + 5*mn;
+  } else {
+    lrwork = std::max(5*mn*mn + 5*mn, 2*mx*mn + 2*mn*mn + mn);
+  }
+  Tensor rwork = at::empty({std::max(int64_t(1), lrwork)}, at::kInt);
+  auto rwork_data = rwork.data_ptr<int>();
+  Tensor iwork = at::empty({8*mn}, at::kInt);
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: Could you please include spaces before and after `*`?
Couple more comments:
1. For LAPACK 3.6, `lrwork` needs to be at least `7 * mn`. Could we do that so that users with older LAPACK versions are not at a disadvantage?
1. Also, in the case `mx > 10 * mn`: how did you obtain `10`?

**Comment by ezyang** [comment]
> Are you able to merge this PR? Thanks.

Sorry, it got stuck in land limbo. Do you want to handle Vishwak's comments?

**Comment by dylanbespalko** [comment]
> > Are you able to merge this PR? Thanks.
> 
> Sorry, it got stuck in land limbo. Do you want to handle Vishwak's comments?

@ezyang, 

Vishwak's concerns are valid, however these values are only impacting the complex number SVD functions because `rwork` is not used for real numbers. If you can merge my changes, I can definitely make these changes in the next PR in a few days. That would help me out a lot.

Dylan.

**Comment by vishwakftw** [comment]
@dylanbespalko thanks for your work, would it be possible for you to address the formatting changes? Also, this needs a rebase.

**Comment by dylanbespalko** [comment]
@vishwakftw,

Yes, can address these issues. I guess there are merge conflicts anyways. I'm busy with other stuff today, but I can have this fixed tomorrow.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 4 above)
Comment (dylanbespalko) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: 1. I have made the changes and update the comment to say it supports LAPACK 3.6.
2. I had to make the assumption that `mx >> mn` implies `mx > 10 * mn`.  I don't support `torch.rand()` with complex numbers so I don't really have a good way to test this right now. 

Once `legacy::cpu::_th_random_` is ported from TH I should be able to better test this. Also, rwork only impacts complex number svd calls.

Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Just one more concern: if rwork is required for complex valued matrices / tensors, why allocate it for the general case?

Can't we special case as follows?
```c++
Tensor rwork;
int* rwork_data = std::nullptr;
if (at::isComplex(self)) {
  rwork = at::empty(.., at::kInt);
  rwork_data = rwork.data_ptr<int>();
}
```

Comment (dylanbespalko) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Good point! I will update the code.

Comment (dylanbespalko) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Ok, I've made the changes. Have a good weekend.

Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Thank you @dylanbespalko, you have a good one too.

**Comment by dylanbespalko** [comment]
@ezyang 

Can you merge these changes?

**Comment by dylanbespalko** [comment]
@ezyang,

I pulled changes from master and re-ran the CI. I don't think these failures are from my code. The previous CI was passing all except: pr/pytorch-win-ws2016-cuda9-cudnn7-py3. That machine is now timing out after 2+ hours.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -978,30 +1089,45 @@ static void apply_svd(Tensor& self, Tensor& U, Tensor& S, Tensor& VT,
   int info;
   auto m = self.size(-2);
   auto n = self.size(-1);
-  auto k = std::min(m, n);
-  Tensor iwork = at::empty({8 * k}, at::kInt);
+  auto mn = std::min(m, n);
+  Tensor iwork = at::empty({8*mn}, at::kInt);
   auto iwork_data = iwork.data_ptr<int>();
+  Tensor rwork;
```
Comment (anjali411) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1095]: @dylanbespalko why is this required? did you have tests to verify the svd behavior for complex?

Comment (anjali411) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1095]: I am working to add svd for complex on cuda if we need to add something similar on cuda as well since magma follows lapack
